File: Napakalaki/src/NapakalakiGame/BadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author Redondo
 */
public class BadConsequence {
    
    private String text;
    private int levels;
    private int nVisibleTreasures;
    private int nHiddenTreasures;
    private boolean death;
    
    public static final int MAXTREASURES = 10 ; 
    

    public BadConsequence(String text, int levels, int nVisibleTreasures, int nHiddenTreasures) {
        this.text = text;
        this.levels = levels;
        this.nVisibleTreasures = nVisibleTreasures;
        this.nHiddenTreasures = nHiddenTreasures;
    }

    public BadConsequence(String text, boolean death) {
        this.text = text;
        this.death = death;
        /* this.levels = Player.MAXLEVEL;
        this.nVisibleTreasures = MAXTREASURES;
        this.nHiddenTreasures = MAXTREASURES; */
    }
    
    private ArrayList<TreasureKind> specificHiddenTreasures = new ArrayList();
    private ArrayList<TreasureKind> specificVisibleTreasures = new ArrayList();

    
    public BadConsequence(String text, int levels, ArrayList<TreasureKind> tVisible, ArrayList<TreasureKind> tHidden) {
            this.text = text;
            this.levels = levels;
            specificVisibleTreasures = tVisible;
            specificHiddenTreasures = tHidden;
    }
    
    
    public String getText() {
        return text;
    }

    public int getLevels() {
        return levels;
    }

    @Override
    public String toString() {
        return "BadConsequence{" + "text=" + text + ", levels=" + levels + ", nVisibleTreasures=" + nVisibleTreasures + ", nHiddenTreasures=" + nHiddenTreasures + ", death=" + death + ", specificHiddenTreasures=" + specificHiddenTreasures + ", specificVisibleTreasures=" + specificVisibleTreasures + '}';
    }

    public int getnVisibleTreasures() {
        return nVisibleTreasures;
    }

    public int getnHiddenTreasures() {
        return nHiddenTreasures;
    }

    public boolean isDeath() {
        return death;
    }

    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
        return specificHiddenTreasures;
    }

    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
        return specificVisibleTreasures;
    }

    public void setText(String text) {
        this.text = text;
    }

    public void setLevels(int levels) {
        this.levels = levels;
    }

    public void setnVisibleTreasures(int nVisibleTreasures) {
        this.nVisibleTreasures = nVisibleTreasures;
    }

    public void setnHiddenTreasures(int nHiddenTreasures) {
        this.nHiddenTreasures = nHiddenTreasures;
    }

    public void setDeath(boolean death) {
        this.death = death;
    }

    public void setSpecificHiddenTreasures(ArrayList<TreasureKind> specificHiddenTreasures) {
        this.specificHiddenTreasures = specificHiddenTreasures;
    }

    public void setSpecificVisibleTreasures(ArrayList<TreasureKind> specificVisibleTreasures) {
        this.specificVisibleTreasures = specificVisibleTreasures;
    }
    
    public boolean  isEmpty(){
        
        if (nVisibleTreasures > 0 )
            return false ; 
        else if(nHiddenTreasures > 0)
            return false ;  
        else if(!specificHiddenTreasures.isEmpty())
            return false ; 
        else if(!specificVisibleTreasures.isEmpty())
            return false ; 
        
        return true ; 
    }
    
    
    public void substractVisibleTreasure(Treasure t){
        
        if(specificVisibleTreasures.contains(t.getType()))
           specificVisibleTreasures.remove(t.getType()) ; 
        else if(nVisibleTreasures > 0)
            nVisibleTreasures-- ; 

    }
    
      public void substractHiddenTreasure(Treasure t){
        
            if(specificHiddenTreasures.contains(t.getType()))
           specificHiddenTreasures.remove(t.getType()) ; 
        else if(nHiddenTreasures > 0)
            nHiddenTreasures-- ; 
      }
    
    public BadConsequence adjustToFitTreasureList(ArrayList<Treasure> v, ArrayList<Treasure> h){
   
        if(death) return(new BadConsequence(this.text, true)) ; 
            
        if (nVisibleTreasures > 0 || nHiddenTreasures > 0) {
            int nv = this.getnVisibleTreasures(); 
            int nh = this.getnHiddenTreasures();
            if(nv > v.size())
                nv = v.size() ; 
            if(nh > h.size())
                nh = h.size() ; 
        
            return (new BadConsequence(this.getText(), this.getLevels(), nv, nh)) ; 
            
        } else {
            
            ArrayList<TreasureKind> nuevovisible = new ArrayList() ; 
            for (TreasureKind tipo: specificVisibleTreasures) {
                for (Treasure tesoro: v)
                    if (tesoro.getType() == tipo)
                        nuevovisible.add(tipo) ; 
                break; 
            }
            
            ArrayList<TreasureKind> nuevooculto = new ArrayList(); 
            for (TreasureKind tipo: specificVisibleTreasures) {
                for (Treasure tesoro: h)
                    if (tesoro.getType() == tipo)
                        nuevooculto.add(tipo); 
                break; 
            }
            
            return (new BadConsequence(this.getText(), this.getLevels(), nuevovisible, nuevooculto)); 
        }
    }
}

File: NapakalakiRuby/pkg/NapakalakiRuby-0.0.1/lib/bad_consequence.rb
# encoding: UTF-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


module NapakalakiGame

require_relative 'treasure_kind'
require_relative 'player'

class BadConsequence
  
  @@MAXTREASURES=10;
  
  private_class_method :new
  def initialize(aText, someLevels, someVisibleTreasures, someHiddenTreasures,
    someSpecificVisibleTreasures, someSpecificHiddenTreasures, death)
    @text = aText
    @levels = someLevels
    @nVisibleTreasures = someVisibleTreasures
    @nHiddenTreasures = someHiddenTreasures
    @death = death 
    @specificVisibleTreasures = someSpecificVisibleTreasures
    @specificHiddenTreasures = someSpecificHiddenTreasures 
  end
  
  attr_accessor :text, :levels, :nVisibleTreasures, :nHiddenTreasures, :death, :specificVisibleTreasures, :specificHiddenTreasures
  
  def to_s
   "Texto: #{@text}\nNiveles perdidos: #{@levels}\n
    Tesoros perdidos visibles: #{@nVisibleTreasures}\n
    Tesoros perdidos ocultos: #{@nHiddenTreasures}\n 
    Death: #{@death}"
  end

  def BadConsequence.getMaxTreasures()
    return @@MAXTREASURES
  end

  def BadConsequence.newLevelNumberOfTreasures (aText, someLevels, someVisibleTreasures, someHiddenTreasures)
   new(aText, someLevels, someVisibleTreasures, someHiddenTreasures, Array.new, Array.new, false)
  end

  def BadConsequence.newLevelSpecificTreasures (aText, someLevels, someSpecificVisibleTreasures, someSpecificHiddenTreasures)
    new(aText, someLevels, 0, 0, someSpecificVisibleTreasures, someSpecificHiddenTreasures, false)
  end 

  def BadConsequence.newDeath (aText) 
    new(aText, Player.getMaxLevel(), @@MAXTREASURES, @@MAXTREASURES, Array.new, Array.new, true)
  end

  def isDeath()
    return @death
  end

  def getSpecificVisibleTreasures()
    return @specificVisibleTreasures;
  end

  def getSpecificHiddenTreasures()
    return @specificHiddenTreasures
  end

  def isEmpty?
    if @nVisibleTreasures == 0 and @nHiddenTreasures == 0 and
        @specificVisibleTreasures.empty? and @specificHiddenTreasures.empty?
      return true
    else 
      return false
    end

    end

  def substractVisibleTreasure(treasure) 
    @specificVisibleTreasures.delete(treasure)
  end

  def substractHiddenTreasure(treasure) 
    @specificHiddenTreasures.delete(treasure)
  end

  def adjustToFitTreasureLists(v, h)
 
    if @death
      return BadConsequence.newDeath(@text)
    end
    
    if @nVisibleTreasures > 0 or @nHiddenTreasures > 0
      nv = @nVisibleTreasures
      nh = @nHiddenTreasures
      
      if nv > v.size
        nv = v.size
      end
      if nh > h.size
        nh = h.size
      end
      
      return BadConsequence.newLevelNumberOfTreasures(@text, @levels, nv, nh)
      
    else 
      
      nuevovisible = Array.new 
      for tipo in @specificVisibleTreasures
        for tesoro in v
          if tesoro.type == tipo
            nuevovisible << new(tipo)
          end
        end
        break
      end
      
      nuevooculto = Array.new
      for tipo in @specificHiddenTreasures
        for tesoro in h
          if tesoro.type == tipo 
            nuevovisible << new(tipo)
          end
        end
        break
      end
      
      return BadConsequence.newLevelSpecificTreasures(@text, @levels, nuevovisible, nuevooculto)
      
    end
  end
  
end 

end